c_src/TLH_absTab.c

```diff
@@ -0,0 +1,225 @@
+         */
+    }
+    /* destroy level1 table */
+    g_hash_table_destroy(tab->lev1);
+    free(tab);
+
+    return 1;
+}
+
+int insert_absTab(AbsTab_t tab, Message *msg)
+{
+    /* IMPORTANT: elements are NOT copied in the
+     * hash tables; only the pointers are stored.
+     */
+    /* msg has three parts:
+     * 1- watch descriptor
+     * 2- length
+     * 3- filename
+     */
+    if(msg == NULL) {
+        return 0;
+    }
+    int ret;
+    GHashTable *lev2_tab = g_hash_table_lookup(tab->lev1, 
+                                               &(msg->wd));
+    if(lev2_tab == NULL) {
+        /* we have a new directory wd. Insert to level1 */
+        /* first create a level2 table for this wd (the value) */
+        lev2_tab = g_hash_table_new(g_str_hash, g_str_equal);
+        /* TODO: check for success */
+        g_hash_table_insert(tab->lev1, &(msg->wd), lev2_tab);
+    }
+    
+    /* add filename to the level2 hash table of this wd */
+    /* we use add rather than insert as we only care about the key (filename) */
+    ret = g_hash_table_add(lev2_tab, msg->name); /* TODO: store len as well */
+    
+    return ret;
+}
+
+int find_absTab(AbsTab_t tab, Message *msg, void *val)
+{
+    (void) val; //To eliminate unused variable warning
+    if(msg == NULL) {
+        return 0;
+    }
+    /* look up wd first */
+    GHashTable *lev2_tab = g_hash_table_lookup(tab->lev1,
+                                               &(msg->wd));
+    /* look up name */
+    if(lev2_tab != NULL) {
+        if(g_hash_table_contains(lev2_tab, msg->name)) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+int del_absTab(AbsTab_t tab, Message *msg)
+{
+    if(msg == NULL) {
+        return 0;
+    }
+    int ret = 1;
+    GHashTable *lev2_tab = g_hash_table_lookup(tab->lev1, 
+                                               &(msg->wd));
+    if(lev2_tab != NULL) {
+        ret = g_hash_table_remove(lev2_tab, msg->name);
+        
+        /* TODO: Free the message memory if not used for wd */
+        
+        /* We should also remove wd from level1 if the corresponding level2
+         * table becomes empty. In such case, the level2 table should be 
+         * destroyed too.
+         */
+        if(g_hash_table_size(lev2_tab) == 0) {
+            g_hash_table_remove(tab->lev1, &(msg->wd));
+            g_hash_table_destroy(lev2_tab);
+        }
+
+        return ret; 
+    }
+
+    return 0;
+}
+
+static int init_rmItr(AbsTab_t tab)
+{
+    g_hash_table_iter_init(&(tab->lev1_rmItr), tab->lev1);
+    if(g_hash_table_iter_next(&(tab->lev1_rmItr), 
+                              (void**) &(tab->lev1_rmItr_key),
+                              (void**) &(tab->lev1_rmItr_val))) {
+        g_hash_table_iter_init(&(tab->lev2_rmItr), tab->lev1_rmItr_val);
+        tab->rmItr_initialized = 1;
+        return 1;
+    }
+
+    tab->rmItr_initialized = 0;
+    return 0;
+}
+
+int remove_next_absTab(AbsTab_t tab, Message *msg)
```

Comment:
For iterators like interface, it's best if you store the state of the iterator in an external object. It avoids sharing data between two usage of the iterator which is a potent source of mistakes. So while it's probably fine in this instance, I think it would make the code a bit cleaner.